sql/src/main/java/io/crate/analyze/where/DocKeys.java

```
@@ -0,0 +1,174 @@
+        }
+
+        public Optional<Long> version() {
+            if (withVersions && key.get(width) != null) {
+                return Optional.of((Long) ((Literal) key.get(width)).value());
+            }
+            return Optional.absent();
+        }
+
+        public String routing() {
+            if (clusteredByIdx >= 0) {
+                return ValueSymbolVisitor.STRING.process(key.get(clusteredByIdx));
+            } else {
+                return id();
+            }
+
+        }
+
+        public Optional<List<BytesRef>> partitionValues() {
+            if (partitionIdx == null || partitionIdx.isEmpty()) {
+                return Optional.absent();
+            }
+            List<BytesRef> values = Lists.transform(partitionIdx, new Function<Integer, BytesRef>() {
+                @Nullable
+                @Override
+                public BytesRef apply(Integer input) {
+                    return ValueSymbolVisitor.BYTES_REF.process(key.get(input));
+                }
+            });
+            return Optional.of(values);
+        }
+
+        public String id() {
+            if (id == null) {
+                if (width == 1) {
+                    id = ValueSymbolVisitor.STRING.process(key.get(0));
+                } else {
+                    BytesStreamOutput out = new BytesStreamOutput(width * 16);
+                    try {
+                        out.writeVInt(width);
+                        if (clusteredByIdx >= 0) {
+                            out.writeBytesRef(ValueSymbolVisitor.BYTES_REF.process(key.get(clusteredByIdx)));
+                        }
+                        for (int i = 0; i < width; i++) {
+                            if (i != clusteredByIdx) {
+                                out.writeBytesRef(ValueSymbolVisitor.BYTES_REF.process(key.get(i)));
+                            }
+                        }
+                        out.close();
+                    } catch (IOException e) {
+                        //
```

maybe bubble the Exception as RuntimeException using ``Throwables.propagate()`` ?